org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTParser.java

```diff
@@ -245,10 +248,13 @@ private List<Classpath> getClasspath() throws IllegalStateException {
 			}
 			if (this.sourcepaths != null) {
 				for (int i = 0, max = this.sourcepaths.length; i < max; i++) {
-					String encoding = this.sourcepathsEncodings == null ? null : this.sourcepathsEncodings[i];
-					main.processPathEntries(
-							Main.DEFAULT_SIZE_CLASSPATH,
-							allClasspaths, this.sourcepaths[i], encoding, true, false);
+					String sourcePath = this.sourcepaths[i];
+					if (sourcePath != null) {
+						String encoding = this.sourcepathsEncodings == null ? null : this.sourcepathsEncodings[i];
```

Comment:
Making the encoding part of the API seems wrong as currently the encoding is not used consistently: Only Batch Compiler uses it in org.eclipse.jdt.internal.compiler.batch.CompilationUnit.getContents() while the IDE uses org.eclipse.jdt.internal.core.util.Util.getResourceContentsAsCharArray(IFile) which disrespect the given classpath encoding but uses the encoding configured in workspace and the utf bom. You may want to harmonize that first.

Reply:
Fixing existing issues is going to make things much harder and longer on this review. We cannot afford fixing all pre-existing issues right now. Can we just open an issue about this `encoding` inconsistency in existing code and look at it later (including changing the new ICompilationUnitResolver interface if necessary)?

Reply:
> Can we just open an issue about this `encoding` inconsistency

sure, just please extend the javadoc of the proposed provisional API with a reference to it so that it gets fixed before publishing such API.

Reply:
I've opened https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2641